pandas/core/reshape/concat.py

```diff
@@ -687,7 +687,23 @@ def _get_concat_axis(self) -> Index:
         if self.keys is None:
             if self.levels is not None:
                 raise ValueError("levels supported only when keys is not None")
-            concat_axis = _concat_indexes(indexes)
+
+            """
+            GH24845 - I have found that the issue only occurs where the categories
+                      in CategoricalIndex are permutations, as shown in the example
+                      on the issue page. The problem also occurs when 'ordered' is
+                      set to False for both CategoricalIndex objects, which led me
+                      to believe that the function cannot differentiate between
+                      the categories. That said, I have devised this solution.
+                      I anticipate that I may need to optimize the code,
+                      but I just wanted to get an idea of what you
+                      (the moderators) think.
+                      https://github.com/pandas-dev/pandas/issues/24845
+            """
+            if check_permutations(indexes):
+                concat_axis = _concat_indexes_permutations(indexes)
+            else:
+                concat_axis = _concat_indexes(indexes)
```

Comment:
You are going around the true problem here. Even though it works, it would be more elegant if you solve the underlying bug. Using your unit test as example (please check the test file):
```
c1.append(c2)
```
returns something not meaningful. If you fix that problem, you consequently fix the DataFrame concatenation problem. I would take a look into the `_concat` method in category: https://github.com/pandas-dev/pandas/blob/efd15b7191c0fea41eb5fc10a8667dd33296d508/pandas/core/indexes/category.py#L572-L584.